### ISSUE TITLE
`useLayoutEffect` handler in redirect should not return anything

### ISSUE THREAD

**Issue by thurinus** [molefrog/wouter#136]
I'm evaluating Wouter for a small Typescript-based React project, and am running into an issue when attempting a `Redirect` while hash-based routing is set up.

I created a [sandbox that replicates the problem](https://codesandbox.io/s/upbeat-knuth-c3pv4), lightly-edited from the example in the readme. This is the error that appears when the redirect link is clicked:

```
TypeError
func.apply is not a function
HTMLUnknownElement.callCallback
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:95:18
Object.invokeGuardedCallbackDev
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:119:20
invokeGuardedCallback
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:150:35
safelyCallDestroy
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:13242:9
eval
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:13590:25
unstable_runWithPriority
https://c3pv4.csb.app/node_modules/scheduler/cjs/scheduler.development.js:653:12
runWithPriority$1
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7219:14
commitUnmount
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:13585:19
unmountHostComponents
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:13848:11
commitDeletion
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:13873:9
commitMutationEffects
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:15294:15
HTMLUnknownElement.callCallback
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:95:18
Object.invokeGuardedCallbackDev
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:119:20
invokeGuardedCallback
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:150:35
commitRootImpl
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:15112:13
unstable_runWithPriority
https://c3pv4.csb.app/node_modules/scheduler/cjs/scheduler.development.js:653:12
runWithPriority$1
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7219:14
commitRoot
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:15033:7
finishSyncRender
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:14655:7
performSyncWorkOnRoot
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:14647:11
eval
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7258:28
unstable_runWithPriority
https://c3pv4.csb.app/node_modules/scheduler/cjs/scheduler.development.js:653:12
runWithPriority$1
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7219:14
flushSyncCallbackQueueImpl
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7254:11
flushSyncCallbackQueue
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:7245:7
scheduleUpdateOnFiber
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:14290:13
dispatchAction
https://c3pv4.csb.app/node_modules/react-dom/cjs/react-dom.development.js:10463:9
handler
/src/App.tsx:15:26
  12 | 
  13 | useEffect(() => {
  14 |   // this function is called whenever the hash changes
> 15 |   const handler = () => setLoc(currentLocation());
     |                        ^
  16 | 
  17 |   // subscribe to hash changes
  18 |   window.addEventListener("hashchange", handler);
```

Apologies in advance if I'm overlooking something basic. Thanks for any help!

**Comment by molefrog** [comment]
@thurinus Hi, thanks for reporting! This appears to be a bug in wouter, I'm going to resolve in an upcoming release.
In the meantime you can fix the issue by making sure `handler()` returns nothing:

```diff
- const navigate = useCallback((to) => (window.location.hash = to), []);
+ const navigate = useCallback((to) => { window.location.hash = to }, []);
```

**Comment by molefrog** [comment]
The regression strategy:
  - [ ] Rewrite `useLayoutEffect` in Redirect component to suppress the `navRef.current()` value to be returned. React thinks it an unsubscribe function and tries to call it, which may lead to an error.
  - [ ] Write a regression spec that covers the behaviour described in this issue.

**Comment by cbbfcd** [comment]
so??
```js
useLayoutEffect(() => { navRef.current() }, []);
```

**Comment by molefrog** [comment]
This has been fixed in `v2.5.2` https://github.com/molefrog/wouter/releases/tag/v2.5.2
